solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStatesOps.java

```diff
@@ -0,0 +1,301 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.common.cloud;
+
+import java.lang.invoke.MethodHandles;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.Op;
+import org.apache.zookeeper.data.Stat;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static java.util.Collections.singletonList;
+
+/**
+ * This is a helper class that encapsulates various operations performed on the per-replica states
+ * Do not directly manipulate the per replica states as it can become difficult to debug them
+ */
+public class PerReplicaStatesOps {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private PerReplicaStates rs;
+  List<PerReplicaStates.Operation> ops;
+  private boolean preOp = true;
+  final Function<PerReplicaStates, List<PerReplicaStates.Operation>> fun;
+
+  PerReplicaStatesOps(Function<PerReplicaStates, List<PerReplicaStates.Operation>> fun) {
+    this.fun = fun;
+  }
+
+  /**
+   * Persist a set of operations to Zookeeper
+   */
+  private void persist(List<PerReplicaStates.Operation> operations, String znode, SolrZkClient zkClient) throws KeeperException, InterruptedException {
+    if (operations == null || operations.isEmpty()) return;
+    if (log.isDebugEnabled()) {
+      log.debug("Per-replica state being persisted for : '{}', ops: {}", znode, operations);
+    }
+
+    List<Op> ops = new ArrayList<>(operations.size());
+    for (PerReplicaStates.Operation op : operations) {
+      //the state of the replica is being updated
+      String path = znode + "/" + op.state.asString;
+      ops.add(op.typ == PerReplicaStates.Operation.Type.ADD ?
+          Op.create(path, null, zkClient.getZkACLProvider().getACLsToAdd(path), CreateMode.PERSISTENT) :
+          Op.delete(path, -1));
+    }
+    try {
+      zkClient.multi(ops, true);
+    } catch (KeeperException e) {
+      log.error("multi op exception : " + e.getMessage() + zkClient.getChildren(znode, null, true));
+      throw e;
+    }
+
+  }
+
+  static List<PerReplicaStates.Operation> addDeleteStaleNodes(List<PerReplicaStates.Operation> ops, PerReplicaStates.State rs) {
```

Comment:
this is kind of awkwardly named, add a java doc or maybe don't call them stale? what makes the nodes stale?